driver-reactive-streams/src/main/com/mongodb/reactivestreams/client/internal/PublisherCreator.java

```diff
@@ -0,0 +1,378 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.mongodb.reactivestreams.client.internal;
+
+import com.mongodb.MongoBulkWriteException;
+import com.mongodb.MongoClientException;
+import com.mongodb.MongoException;
+import com.mongodb.MongoNamespace;
+import com.mongodb.MongoWriteConcernException;
+import com.mongodb.MongoWriteException;
+import com.mongodb.ReadConcern;
+import com.mongodb.ReadPreference;
+import com.mongodb.WriteConcern;
+import com.mongodb.WriteConcernResult;
+import com.mongodb.WriteError;
+import com.mongodb.bulk.BulkWriteResult;
+import com.mongodb.client.model.CreateCollectionOptions;
+import com.mongodb.client.model.CreateViewOptions;
+import com.mongodb.client.model.IndexModel;
+import com.mongodb.client.model.IndexOptionDefaults;
+import com.mongodb.client.model.ValidationOptions;
+import com.mongodb.internal.async.SingleResultCallback;
+import com.mongodb.internal.async.client.AsyncClientSession;
+import com.mongodb.internal.async.client.OperationExecutor;
+import com.mongodb.internal.bulk.WriteRequest;
+import com.mongodb.internal.client.model.AggregationLevel;
+import com.mongodb.internal.client.model.changestream.ChangeStreamLevel;
+import com.mongodb.internal.operation.AsyncReadOperation;
+import com.mongodb.internal.operation.AsyncWriteOperation;
+import com.mongodb.internal.operation.CommandReadOperation;
+import com.mongodb.internal.operation.CreateCollectionOperation;
+import com.mongodb.internal.operation.CreateViewOperation;
+import com.mongodb.internal.operation.DropDatabaseOperation;
+import com.mongodb.internal.operation.IndexHelper;
+import com.mongodb.lang.Nullable;
+import com.mongodb.reactivestreams.client.AggregatePublisher;
+import com.mongodb.reactivestreams.client.ChangeStreamPublisher;
+import com.mongodb.reactivestreams.client.ClientSession;
+import com.mongodb.reactivestreams.client.DistinctPublisher;
+import com.mongodb.reactivestreams.client.FindPublisher;
+import com.mongodb.reactivestreams.client.ListCollectionsPublisher;
+import com.mongodb.reactivestreams.client.ListDatabasesPublisher;
+import com.mongodb.reactivestreams.client.ListIndexesPublisher;
+import com.mongodb.reactivestreams.client.MapReducePublisher;
+import org.bson.BsonDocument;
+import org.bson.codecs.configuration.CodecRegistry;
+import org.bson.conversions.Bson;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.core.publisher.MonoSink;
+
+import java.util.List;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import static com.mongodb.assertions.Assertions.notNull;
+import static java.util.stream.Collectors.toList;
+
+public final class PublisherCreator {
```

Comment:
If we do end up leaving this as purely static, there's a cool trick I just learned: you can make this an interface instead of a final class

Reply:
Refactored this into MongoOperationsPublisher - which is a bad name but that now handles all client, database, collection operations.